src/mdns.c

```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
                 return (MDNS_ERROR);
         }
         *pp_intfs = intfs = malloc(sizeof(*intfs) * nb_if);
-        if (intfs == NULL) {
-                freeifaddrs(ifs);
-                return (MDNS_ERROR);
-        }
+        if (intfs == NULL) 
+                goto error;
+
+        *pp_ip_addresses = ip_addresses = malloc(sizeof(char*) * nb_if);
+
+        if (ip_addresses == NULL) 
+                goto error;
+
         for (c = ifs; c != NULL; c = c->ifa_next) {
                 if (c->ifa_addr == NULL ||
                     c->ifa_addr->sa_family != ai_family ||
                     !mdns_is_interface_valuable(c))
                         continue;
                 memcpy(intfs, c->ifa_addr, sizeof(*intfs));
+                
+                char address_buf[INET6_ADDRSTRLEN];
+
+                const char *ip_address;
+                if (ai_family == AF_INET)
+                {
+                        struct sockaddr_in *addr_in = (struct sockaddr_in *)c->ifa_addr;
+                        ip_address = inet_ntop( ai_family, &addr_in->sin_addr, 
+                                                    address_buf, sizeof( address_buf ) );
+                }
+                else
+                {
+                        struct sockaddr_in6 *addr_in = (struct sockaddr_in6 *)c->ifa_addr;
+                        ip_address = inet_ntop( ai_family, &addr_in->sin6_addr, 
+                                                    address_buf, sizeof( address_buf ) );
+                }
+
+                *ip_addresses = strdup(ip_address);
+                if (*ip_addresses == NULL)
+                        goto error;
+                ip_addresses++;
                 intfs++;
         }
         freeifaddrs(ifs);
         *p_nb_intf = nb_if;
         return (0);
+error:
+        while (ip_addresses != *pp_ip_addresses)
+        {
+                free(*ip_addresses);
+                ip_addresses--;
+        }
+        free(*pp_ip_addresses);
+        free(*pp_intfs);
+        freeifaddrs(ifs);
+        return (MDNS_ERROR);
 }
 #else
 static size_t
-mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
+mdns_list_interfaces(multicast_if** pp_intfs, char ***pp_ip_addresses, size_t* p_nb_intf, int ai_family)
 {
         multicast_if *intfs;
         *pp_intfs = intfs = malloc(sizeof(*intfs));
         if (intfs == NULL)
                 return (MDNS_ERROR);
         memset(intfs, 0, sizeof(*intfs));
+
+        *pp_ip_addresses = malloc(sizeof(char*));
+        if (pp_ip_addresses == NULL)
```

Comment:
You're not checking the pointer that was allocated

Reply:
Oops - I'll fix this. Thanks.